webapp/lib/controller.dart

```diff
@@ -275,38 +329,38 @@ void fetchConfigurationData() {
   configurationResponseLanguages = configurationReponseLanguageData;
 }
 
-Map<String, bool> getTagList(String selectedTag, Map<String, Map<String, List<String>>> tagData) {
+Map<String, bool> getTagList(String selectedTag, Map<String, List<List<String>>> tagData) {
   return new Map.fromIterable(tagData.keys.toList(),
     key: (tag) => tag,
     value: (tag) => selectedTag != null && selectedTag == tag ? true : false);
 }
 
-void populateConfigurationView(String selectedTag, Map<String, bool> tagList, Map<String, List<String>> tagResponses) {
+void populateConfigurationView(String selectedTag, Map<String, bool> tagList, List<String> responseLanguages, List<List<String>> tagResponses) {
   view.contentView.configurationView.tagList.renderTagList(tagList);
-  view.contentView.configurationView.tagResponses.renderResponses(selectedTag, tagResponses);
+  view.contentView.configurationView.tagResponses.renderResponses(selectedTag, responseLanguages, tagResponses);
 }
 
-void addNewConfigurationTag(String tagToAdd, List<String> availableLanguages, Set<String> additionalTags, Map<String, Map<String, List<String>>> tagData) {
-  tagData[tagToAdd] = new Map.fromIterable(availableLanguages, key: (d) => d, value: (d) => ['']);
+void addNewConfigurationTag(String tagToAdd, List<String> availableLanguages, Set<String> additionalTags, Map<String, List<List<String>>> tagData) {
+  tagData[tagToAdd] = [['','']];
   additionalTags.remove(tagToAdd);
-  populateConfigurationView(tagToAdd, getTagList(tagToAdd, tagData), tagData[tagToAdd]);
+  populateConfigurationView(tagToAdd, getTagList(tagToAdd, tagData), availableLanguages, tagData[tagToAdd]);
 }
 
-void updateEditedConfigurationTagResponse(String parentTag, int index, String language, String text) {
-  configurationTagData[parentTag][language][index]= text;
+void updateEditedConfigurationTagResponse(String parentTag, int textIndex, int languageIndex, String text, List<List<String>> tagResponses) {
+  tagResponses[textIndex][languageIndex] = text;
 }
 
-void addConfigurationResponseEntries(String parentTag, String language, String text, Map<String, Map<String, List<String>>> tagData) {
-  if (language != null && text != null) {
-    var pos = tagData[parentTag][language].indexOf('');
+void addConfigurationResponseEntries(String parentTag, int languageIndex, String text, List<String> responseLanguages, Map<String, List<List<String>>> tagData) {
+  if (languageIndex != null && text != null) {
+    var pos = tagData[parentTag].indexWhere((x)=> x.contains(''));
     if (pos > -1) {
-      tagData[parentTag][language][pos] = text;
+      tagData[parentTag][pos][languageIndex] = text;
     } else {
-      tagData[parentTag].forEach((k, v) => v.add(''));
-      tagData[parentTag][language].last = text;
+      tagData[parentTag].add(['', '']);
```

Comment:
same as above - I think `responseLanguages.map((e) => '').toList()` should do it.

Reply:
I agree. Refactored it.